File: quant1x/level1/heartbeat.rs
use super::sequence_id;
use super::BinaryStream;
use crate::level1::commands::*;
use crate::level1::protocol::{Request, RequestHeader, Response, ResponseHeader};

#[derive(Debug, Clone)]
pub struct HeartbeatRequest {
    pub header: RequestHeader,
}

impl Request for HeartbeatRequest {
    fn header(&self) -> &RequestHeader {
        &self.header
    }

    fn header_mut(&mut self) -> &mut RequestHeader {
        &mut self.header
    }

    fn serialize_payload(&mut self) -> Vec<u8> {
        self.header.pkg_len1 = 2;
        self.header.pkg_len2 = 2;
        BinaryStream::new().data().clone()
    }

    fn payload_string(&self) -> String {
        "HeartbeatRequest".to_string()
    }
}

impl HeartbeatRequest {
    pub fn new() -> Self {
        let mut header = RequestHeader::new();
        header.zip_flag = crate::level1::protocol::zlib_flag::UNCOMPRESSED;
        header.seq_id = sequence_id();
        header.packet_type = 0x02;
        header.method = HEARTBEAT;

        HeartbeatRequest { header }
    }
}

#[derive(Debug, Clone)]
pub struct HeartbeatResponse {
    pub header: ResponseHeader,
    pub info: String,
}

impl Response for HeartbeatResponse {
    fn header(&self) -> &ResponseHeader {
        &self.header
    }

    fn header_mut(&mut self) -> &mut ResponseHeader {
        &mut self.header
    }

    fn deserialize_body(&mut self, data: &[u8]) -> Result<(), crate::std::DeserializeError> {
        let mut bs = BinaryStream::from_vec(data.to_vec());
        // The original code read 10 bytes string?
        // "self.info = bs.get_string(10)?;"
        // Assuming the body contains this.
        self.info = bs.get_string(10)?;
        Ok(())
    }

    fn body_string(&self) -> String {
        format!("Info: {}", self.info)
    }
}

impl HeartbeatResponse {
    pub fn new() -> Self {
        Self {
            header: ResponseHeader::new(),
            info: String::new(),
        }
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    #[test]
    fn deserialize_sample_populates_info() {
        let hex_data = "48656172742d486562696f726974";
        let buf = hex::decode(hex_data).unwrap();
        let mut resp = HeartbeatResponse::new();
        resp.deserialize(&buf).unwrap();
        assert!(!resp.info.is_empty());
    }
}

File: quant1x/level1/heartbeat.py
# -*- coding: UTF-8 -*-
# Copyright (c) Quant1X <wangfengxy@sina.cn>.
# Licensed under the MIT License.

from __future__ import annotations

import struct
from quant1x.level1.protocol import (
    FLAG_UNCOMPRESSED,
    COMMAND_HEARTBEAT,
    sequence_id,
)

class HeartbeatRequest:
    """
    心跳请求
    """
    def __init__(self):
        self.zip_flag = FLAG_UNCOMPRESSED
        self.seq_id = sequence_id()
        self.packet_type = 0x02
        self.body_wire_len = 0
        self.body_raw_len = 0
        self.method = COMMAND_HEARTBEAT

    def serialize(self) -> bytes:
        self.body_wire_len = 2
        self.body_raw_len = 2
        
        header = struct.pack('<B I B H H H', self.zip_flag, self.seq_id, self.packet_type, self.body_wire_len, self.body_raw_len, self.method)
        return header

class HeartbeatResponse:
    """
    心跳响应
    """
    def __init__(self):
        self.info = ""

    def deserialize(self, data: bytes) -> None:
        # get_string(10)
        if len(data) >= 10:
            s = data[:10]
        else:
            s = data
            
        try:
            self.info = s.decode('gbk', errors='ignore').split('\x00', 1)[0]
        except Exception:
            self.info = s.decode('utf-8', errors='ignore').split('\x00', 1)[0]
